File: app/src/main/java/com/oreki/stumpd/utils/RankingUtils.kt
package com.oreki.stumpd.utils

import com.oreki.stumpd.data.manager.MatchPerformance
import com.oreki.stumpd.data.manager.PlayerDetailedStats
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow

/**
 * ICC-inspired player ranking system.
 *
 * Key design principles:
 * - Per-match ratings on a 0-1000 scale, **format-aware** (5-over, 10-over, 20-over)
 * - Real calendar-based time decay (weekly, not index-based)
 * - Inactivity penalty: players who haven't played recently see ratings drop
 * - Missed match dilution: players missing group matches get phantom 0-rating entries
 * - Form multiplier: recent 3-match form boosts/dampens career rating by up to ±15%
 * - Per-role experience ramp: need 10 innings in a discipline for full credit
 * - Role-aware overall rating (pure batters aren't penalized for not bowling)
 * - Maiden over bonus for bowling (scales with format)
 */
object RankingUtils {

    private const val WEEKLY_DECAY = 0.97              // Each week of age loses 3% weight
    private const val INACTIVITY_DECAY_PER_WEEK = 0.98 // 2% penalty per inactive week
    private const val INACTIVITY_GRACE_WEEKS = 2.0     // No penalty for first 2 weeks
    private const val EXPERIENCE_RAMP = 10.0           // Innings needed for full rating
    private const val MILLIS_PER_WEEK = 7.0 * 24 * 60 * 60 * 1000
    private const val FORM_MATCHES = 3                 // Number of recent matches for form calc
    private const val MAX_FORM_BOOST = 1.15            // Max +15% for hot streak
    private const val MIN_FORM_PENALTY = 0.85          // Max -15% for cold streak

    // --- Per-innings rating functions (0-1000 scale, format-aware) ---

    /**
     * Rate a single batting innings on a 0-1000 scale.
     *
     * Format-aware calibration:
     *   runsMultiplier = 400 / totalOvers  (80 for 5-over, 40 for 10-over, 20 for 20-over)
     *   srBenchmark = 100 + 500 / totalOvers  (200 for 5-over, 150 for 10-over, 125 for 20-over)
     *
     * 5-over examples (multiplier=80, srBenchmark=200):
     *   5 runs, SR 125 → 250  |  8 runs, SR 200 → 640  |  12 runs, SR 200 → 960
     *
     * 20-over examples (multiplier=20, srBenchmark=125):
     *   30 runs, SR 120 → 576  |  50 runs, SR 143 → 1000
     */
    fun rateBattingInnings(perf: MatchPerformance): Double {
        if (perf.ballsFaced == 0 && perf.runs == 0) return 0.0

        val totalOvers = perf.matchTotalOvers.coerceAtLeast(1)
        val runsMultiplier = 400.0 / totalOvers
        val srBenchmark = 100.0 + (500.0 / totalOvers)
        val notOutThreshold = max(3, totalOvers)

        val basePoints = perf.runs * runsMultiplier
        val strikeRate = if (perf.ballsFaced > 0) {
            (perf.runs.toDouble() / perf.ballsFaced) * 100.0
        } else {
            0.0
        }
        val srFactor = (strikeRate / srBenchmark).coerceIn(0.6, 1.5)
        val notOutBonus = if (!perf.isOut && perf.runs >= notOutThreshold) 50.0 else 0.0
        val raw = basePoints * srFactor + notOutBonus
        return min(raw, 1000.0)
    }

    /**
     * Rate a single bowling innings on a 0-1000 scale.
     *
     * Format-aware calibration:
     *   normalizedOvers = oversBowled * (20 / totalOvers)
     *   -- makes 1 over in a 5-over match equivalent to 4 overs in T20
     *
     * Maiden bonus (scales with format):
     *   maidenBonus = maidens * 100 * (20 / totalOvers)
     *   -- 1 maiden in 5-over match = 400pts (extraordinary!)
     *   -- 1 maiden in 20-over match = 100pts
     *
     * Economy benchmark stays at 7.0 (runs-per-over, format-independent).
     *
     * 5-over examples (1 over, normalizedOvers=4):
     *   0w eco 5.0 → 150  |  maiden 0w → 550  |  2w eco 4.0 → 780  |  4w eco 1.0 → 1000
     */
    fun rateBowlingInnings(perf: MatchPerformance): Double {
        if (perf.ballsBowled == 0) return 0.0

        val totalOvers = perf.matchTotalOvers.coerceAtLeast(1)
        val formatScale = 20.0 / totalOvers

        val oversBowled = perf.ballsBowled / 6.0
        val normalizedOvers = oversBowled * formatScale

        val economy = if (perf.ballsBowled > 0) {
            (perf.runsConceded.toDouble() / perf.ballsBowled) * 6.0
        } else {
            0.0
        }
        val econFactor = if (economy > 0) {
            (7.0 / economy).coerceIn(0.5, 2.0)
        } else {
            1.5
        }

        // Wicket reward (primary) — scaled by format
        val wicketRating = perf.wickets * (180.0 * formatScale / 4.0) * econFactor
        // ↑ At formatScale=4 (5-over), this equals wickets * 180 * econFactor (same as before)

        // Economy base: tight bowling is valuable even without wickets
        val econBase = min(normalizedOvers * 30.0 * econFactor, 150.0)

        // Maiden bonus
        val maidenBonus = perf.maidenOvers * (100.0 * formatScale)

        return min(wicketRating + econBase + maidenBonus, 1000.0)
    }

    /**
     * Rate fielding in a single match on a 0-1000 scale.
     *
     * Formula:
     *   points = catches * 200 + runOuts * 250 + stumpings * 250
     *   rating = min(points, 1000)
     */
    fun rateFieldingMatch(perf: MatchPerformance): Double {
        val points = (perf.catches * 200.0) + (perf.runOuts * 250.0) + (perf.stumpings * 250.0)
        return min(points, 1000.0)
    }

    // --- Career rating functions (time-weighted averages + missed matches + form) ---

    /**
     * Calculate career batting rating (0-1000).
     *
     * @param performances All match performances for the player
     * @param allGroupMatches Optional list of (matchId, matchDate) for all group matches.
     *   When provided, missed matches inject phantom 0-rating entries to dilute the average.
     *   Pass null for "All Groups" view where missed-match penalty doesn't apply.
     */
    fun calculateBattingRating(
        performances: List<MatchPerformance>,
        allGroupMatches: List<Pair<String, Long>>? = null
    ): Double {
        val battingInnings = performances.filter { it.ballsFaced > 0 || it.runs > 0 }
        if (battingInnings.isEmpty()) return 0.0

        val sorted = battingInnings.sortedByDescending { it.matchDate }
        val playedMatchIds = sorted.map { it.matchId }.toSet()

        val rawRating = timeWeightedAverage(sorted, allGroupMatches, playedMatchIds) { rateBattingInnings(it) }
        val experienceFactor = min(battingInnings.size / EXPERIENCE_RAMP, 1.0)
        val inactivityFactor = calculateInactivityFactor(sorted.first().matchDate)
        val formMultiplier = calculateFormMultiplier(sorted) { rateBattingInnings(it) }

        return (rawRating * experienceFactor * inactivityFactor * formMultiplier).coerceIn(0.0, 1000.0)
    }

    /**
     * Calculate career bowling rating (0-1000).
     *
     * @param performances All match performances for the player
     * @param allGroupMatches Optional list of (matchId, matchDate) for all group matches.
     */
    fun calculateBowlingRating(
        performances: List<MatchPerformance>,
        allGroupMatches: List<Pair<String, Long>>? = null
    ): Double {
        val bowlingInnings = performances.filter { it.ballsBowled > 0 }
        if (bowlingInnings.isEmpty()) return 0.0

        val sorted = bowlingInnings.sortedByDescending { it.matchDate }
        val playedMatchIds = sorted.map { it.matchId }.toSet()

        val rawRating = timeWeightedAverage(sorted, allGroupMatches, playedMatchIds) { rateBowlingInnings(it) }
        val experienceFactor = min(bowlingInnings.size / EXPERIENCE_RAMP, 1.0)
        val inactivityFactor = calculateInactivityFactor(sorted.first().matchDate)
        val formMultiplier = calculateFormMultiplier(sorted) { rateBowlingInnings(it) }

        return (rawRating * experienceFactor * inactivityFactor * formMultiplier).coerceIn(0.0, 1000.0)
    }

    /**
     * Calculate career fielding rating (0-1000).
     */
    fun calculateFieldingRating(performances: List<MatchPerformance>): Double {
        val fieldingMatches = performances.filter {
            it.catches > 0 || it.runOuts > 0 || it.stumpings > 0
        }
        if (fieldingMatches.isEmpty()) return 0.0

        val sorted = fieldingMatches.sortedByDescending { it.matchDate }
        val rawRating = timeWeightedAverage(sorted) { rateFieldingMatch(it) }
        val experienceFactor = min(fieldingMatches.size / EXPERIENCE_RAMP, 1.0)
        val inactivityFactor = calculateInactivityFactor(sorted.first().matchDate)
        return rawRating * experienceFactor * inactivityFactor
    }

    /**
     * Calculate overall player rating (0-1000) with role-aware weighting.
     *
     * - All-rounder (batted AND bowled): 40% batting + 40% bowling + 20% fielding
     * - Pure batter (never bowled):      75% batting + 25% fielding
     * - Pure bowler (never batted):      75% bowling + 25% fielding
     *
     * Requires at least 3 matches to qualify.
     *
     * @param allGroupMatches Optional group match context for missed-match penalty
     */
    fun calculateOverallRating(
        player: PlayerDetailedStats,
        allGroupMatches: List<Pair<String, Long>>? = null
    ): Double {
        if (player.matchPerformances.size < 3) return 0.0

        val batting = calculateBattingRating(player.matchPerformances, allGroupMatches)
        val bowling = calculateBowlingRating(player.matchPerformances, allGroupMatches)
        val fielding = calculateFieldingRating(player.matchPerformances)

        val hasBatted = player.totalRuns > 0 || player.totalBallsFaced > 0
        val hasBowled = player.totalBallsBowled > 0

        return when {
            hasBatted && hasBowled -> batting * 0.40 + bowling * 0.40 + fielding * 0.20
            hasBatted -> batting * 0.75 + fielding * 0.25
            hasBowled -> bowling * 0.75 + fielding * 0.25
            else -> fielding
        }
    }

    // --- Legacy compatibility ---

    @Suppress("UNUSED_PARAMETER")
    fun calculateOverallScore(player: PlayerDetailedStats, allPlayers: List<PlayerDetailedStats>): Double {
        return calculateOverallRating(player)
    }

    // --- Helpers ---

    /**
     * Compute a time-weighted average using real calendar decay.
     *
     * Each match is weighted by 0.97^(weeks since match).
     * A match played today has weight ~1.0, a match from 3 months ago ~0.67,
     * a match from 6 months ago ~0.45, a match from 1 year ago ~0.21.
     *
     * When [allGroupMatches] is provided, missed matches (in the group but not played
     * by this player) inject phantom 0-rating entries, diluting the average.
     */
    private fun timeWeightedAverage(
        sortedPerformances: List<MatchPerformance>,
        allGroupMatches: List<Pair<String, Long>>? = null,
        playedMatchIds: Set<String>? = null,
        ratingFn: (MatchPerformance) -> Double
    ): Double {
        val now = System.currentTimeMillis()
        var weightedSum = 0.0
        var totalWeight = 0.0

        // Real performances
        sortedPerformances.forEach { perf ->
            val weeksSince = (now - perf.matchDate).toDouble() / MILLIS_PER_WEEK
            val weight = WEEKLY_DECAY.pow(weeksSince)
            weightedSum += ratingFn(perf) * weight
            totalWeight += weight
        }

        // Phantom 0-rating entries for missed group matches
        if (allGroupMatches != null && playedMatchIds != null) {
            allGroupMatches.forEach { (matchId, matchDate) ->
                if (matchId !in playedMatchIds) {
                    val weeksSince = (now - matchDate).toDouble() / MILLIS_PER_WEEK
                    val weight = WEEKLY_DECAY.pow(weeksSince)
                    // weightedSum += 0 (phantom zero), but totalWeight increases
                    totalWeight += weight
                }
            }
        }

        return if (totalWeight > 0) weightedSum / totalWeight else 0.0
    }

    /**
     * Form multiplier: compare last [FORM_MATCHES] match ratings vs career average.
     *
     * - Hot streak (recent >> career): up to +15%
     * - Cold streak (recent << career): up to -15%
     * - Neutral: 1.0 (no change)
     *
     * Only applies when the player has more than [FORM_MATCHES] innings.
     */
    private fun calculateFormMultiplier(
        sortedPerformances: List<MatchPerformance>,
        ratingFn: (MatchPerformance) -> Double
    ): Double {
        if (sortedPerformances.size <= FORM_MATCHES) return 1.0

        val recentRatings = sortedPerformances.take(FORM_MATCHES).map { ratingFn(it) }
        val careerRatings = sortedPerformances.map { ratingFn(it) }

        val recentAvg = recentRatings.average()
        val careerAvg = careerRatings.average()

        if (careerAvg <= 0) return 1.0

        return (recentAvg / careerAvg).coerceIn(MIN_FORM_PENALTY, MAX_FORM_BOOST)
    }

    /**
     * Inactivity penalty: if a player's most recent match was more than
     * [INACTIVITY_GRACE_WEEKS] ago, their rating decays by 2% per additional week.
     *
     * - Played within 2 weeks: no penalty (factor = 1.0)
     * - 1 month inactive:  factor ≈ 0.96
     * - 3 months inactive: factor ≈ 0.80
     * - 6 months inactive: factor ≈ 0.62
     * - 1 year inactive:   factor ≈ 0.37
     */
    private fun calculateInactivityFactor(mostRecentMatchDate: Long): Double {
        val weeksSinceLast = (System.currentTimeMillis() - mostRecentMatchDate).toDouble() / MILLIS_PER_WEEK
        val inactiveWeeks = (weeksSinceLast - INACTIVITY_GRACE_WEEKS).coerceAtLeast(0.0)
        return INACTIVITY_DECAY_PER_WEEK.pow(inactiveWeeks)
    }
}

File: web/src/utils/rankingUtils.ts
/**
 * ICC-inspired player ranking system.
 * Ported from RankingUtils.kt (Android app).
 *
 * Key design principles:
 * - Per-match ratings on 0-1000 scale, format-aware (5-over, 10-over, 20-over)
 * - Real calendar-based time decay (weekly, not index-based)
 * - Inactivity penalty for players who haven't played recently
 * - Missed match dilution: phantom 0-rating entries for missed group matches
 * - Form multiplier: recent 3-match form boosts/dampens rating by up to ±15%
 * - Per-role experience ramp: 10 innings needed for full credit
 * - Role-aware overall rating (pure batters aren't penalized for not bowling)
 * - Maiden over bonus for bowling (scales with format)
 */

import type { MatchPerformance, PlayerDetailedStats } from '../types';

const WEEKLY_DECAY = 0.97;
const INACTIVITY_DECAY_PER_WEEK = 0.98;
const INACTIVITY_GRACE_WEEKS = 2.0;
const EXPERIENCE_RAMP = 10.0;
const MILLIS_PER_WEEK = 7.0 * 24 * 60 * 60 * 1000;
const FORM_MATCHES = 3;
const MAX_FORM_BOOST = 1.15;
const MIN_FORM_PENALTY = 0.85;

function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, value));
}

// ==================== Per-innings rating functions ====================

export function rateBattingInnings(perf: MatchPerformance): number {
  if (perf.ballsFaced === 0 && perf.runs === 0) return 0;

  const totalOvers = Math.max(perf.matchTotalOvers, 1);
  const runsMultiplier = 400.0 / totalOvers;
  const srBenchmark = 100.0 + 500.0 / totalOvers;
  const notOutThreshold = Math.max(3, totalOvers);

  const basePoints = perf.runs * runsMultiplier;
  const strikeRate =
    perf.ballsFaced > 0 ? (perf.runs / perf.ballsFaced) * 100.0 : 0;
  const srFactor = clamp(strikeRate / srBenchmark, 0.6, 1.5);
  const notOutBonus =
    !perf.isOut && perf.runs >= notOutThreshold ? 50.0 : 0.0;
  const raw = basePoints * srFactor + notOutBonus;
  return Math.min(raw, 1000.0);
}

export function rateBowlingInnings(perf: MatchPerformance): number {
  if (perf.ballsBowled === 0) return 0;

  const totalOvers = Math.max(perf.matchTotalOvers, 1);
  const formatScale = 20.0 / totalOvers;
  const oversBowled = perf.ballsBowled / 6.0;
  const normalizedOvers = oversBowled * formatScale;

  const economy =
    perf.ballsBowled > 0
      ? (perf.runsConceded / perf.ballsBowled) * 6.0
      : 0;
  const econFactor =
    economy > 0 ? clamp(7.0 / economy, 0.5, 2.0) : 1.5;

  const wicketRating =
    perf.wickets * ((180.0 * formatScale) / 4.0) * econFactor;
  const econBase = Math.min(normalizedOvers * 30.0 * econFactor, 150.0);
  const maidenBonus = perf.maidenOvers * 100.0 * formatScale;

  return Math.min(wicketRating + econBase + maidenBonus, 1000.0);
}

export function rateFieldingMatch(perf: MatchPerformance): number {
  const points =
    perf.catches * 200.0 + perf.runOuts * 250.0 + perf.stumpings * 250.0;
  return Math.min(points, 1000.0);
}

// ==================== Career rating functions ====================

function timeWeightedAverage(
  sortedPerformances: MatchPerformance[],
  allGroupMatches: [string, number][] | null,
  playedMatchIds: Set<string> | null,
  ratingFn: (p: MatchPerformance) => number
): number {
  const now = Date.now();
  let weightedSum = 0;
  let totalWeight = 0;

  for (const perf of sortedPerformances) {
    const weeksSince = (now - perf.matchDate) / MILLIS_PER_WEEK;
    const weight = Math.pow(WEEKLY_DECAY, weeksSince);
    weightedSum += ratingFn(perf) * weight;
    totalWeight += weight;
  }

  // Phantom 0-rating entries for missed group matches
  if (allGroupMatches && playedMatchIds) {
    for (const [matchId, matchDate] of allGroupMatches) {
      if (!playedMatchIds.has(matchId)) {
        const weeksSince = (now - matchDate) / MILLIS_PER_WEEK;
        const weight = Math.pow(WEEKLY_DECAY, weeksSince);
        totalWeight += weight;
      }
    }
  }

  return totalWeight > 0 ? weightedSum / totalWeight : 0;
}

function calculateFormMultiplier(
  sortedPerformances: MatchPerformance[],
  ratingFn: (p: MatchPerformance) => number
): number {
  if (sortedPerformances.length <= FORM_MATCHES) return 1.0;

  const recentRatings = sortedPerformances
    .slice(0, FORM_MATCHES)
    .map(ratingFn);
  const careerRatings = sortedPerformances.map(ratingFn);

  const recentAvg =
    recentRatings.reduce((a, b) => a + b, 0) / recentRatings.length;
  const careerAvg =
    careerRatings.reduce((a, b) => a + b, 0) / careerRatings.length;

  if (careerAvg <= 0) return 1.0;
  return clamp(recentAvg / careerAvg, MIN_FORM_PENALTY, MAX_FORM_BOOST);
}

function calculateInactivityFactor(mostRecentMatchDate: number): number {
  const weeksSinceLast =
    (Date.now() - mostRecentMatchDate) / MILLIS_PER_WEEK;
  const inactiveWeeks = Math.max(weeksSinceLast - INACTIVITY_GRACE_WEEKS, 0);
  return Math.pow(INACTIVITY_DECAY_PER_WEEK, inactiveWeeks);
}

export function calculateBattingRating(
  performances: MatchPerformance[],
  allGroupMatches: [string, number][] | null = null
): number {
  const battingInnings = performances.filter(
    (p) => p.ballsFaced > 0 || p.runs > 0
  );
  if (battingInnings.length === 0) return 0;

  const sorted = [...battingInnings].sort(
    (a, b) => b.matchDate - a.matchDate
  );
  const playedMatchIds = new Set(sorted.map((p) => p.matchId));

  const rawRating = timeWeightedAverage(
    sorted,
    allGroupMatches,
    playedMatchIds,
    rateBattingInnings
  );
  const experienceFactor = Math.min(
    battingInnings.length / EXPERIENCE_RAMP,
    1.0
  );
  const inactivityFactor = calculateInactivityFactor(sorted[0].matchDate);
  const formMultiplier = calculateFormMultiplier(sorted, rateBattingInnings);

  return clamp(
    rawRating * experienceFactor * inactivityFactor * formMultiplier,
    0,
    1000
  );
}

export function calculateBowlingRating(
  performances: MatchPerformance[],
  allGroupMatches: [string, number][] | null = null
): number {
  const bowlingInnings = performances.filter((p) => p.ballsBowled > 0);
  if (bowlingInnings.length === 0) return 0;

  const sorted = [...bowlingInnings].sort(
    (a, b) => b.matchDate - a.matchDate
  );
  const playedMatchIds = new Set(sorted.map((p) => p.matchId));

  const rawRating = timeWeightedAverage(
    sorted,
    allGroupMatches,
    playedMatchIds,
    rateBowlingInnings
  );
  const experienceFactor = Math.min(
    bowlingInnings.length / EXPERIENCE_RAMP,
    1.0
  );
  const inactivityFactor = calculateInactivityFactor(sorted[0].matchDate);
  const formMultiplier = calculateFormMultiplier(sorted, rateBowlingInnings);

  return clamp(
    rawRating * experienceFactor * inactivityFactor * formMultiplier,
    0,
    1000
  );
}

export function calculateFieldingRating(
  performances: MatchPerformance[]
): number {
  const fieldingMatches = performances.filter(
    (p) => p.catches > 0 || p.runOuts > 0 || p.stumpings > 0
  );
  if (fieldingMatches.length === 0) return 0;

  const sorted = [...fieldingMatches].sort(
    (a, b) => b.matchDate - a.matchDate
  );
  const rawRating = timeWeightedAverage(sorted, null, null, rateFieldingMatch);
  const experienceFactor = Math.min(
    fieldingMatches.length / EXPERIENCE_RAMP,
    1.0
  );
  const inactivityFactor = calculateInactivityFactor(sorted[0].matchDate);
  return rawRating * experienceFactor * inactivityFactor;
}

export function calculateOverallRating(
  player: PlayerDetailedStats,
  allGroupMatches: [string, number][] | null = null
): number {
  if (player.matchPerformances.length < 3) return 0;

  const batting = calculateBattingRating(
    player.matchPerformances,
    allGroupMatches
  );
  const bowling = calculateBowlingRating(
    player.matchPerformances,
    allGroupMatches
  );
  const fielding = calculateFieldingRating(player.matchPerformances);

  const hasBatted =
    player.totalRuns > 0 || player.totalBallsFaced > 0;
  const hasBowled = player.totalBallsBowled > 0;

  if (hasBatted && hasBowled) {
    return batting * 0.4 + bowling * 0.4 + fielding * 0.2;
  } else if (hasBatted) {
    return batting * 0.75 + fielding * 0.25;
  } else if (hasBowled) {
    return bowling * 0.75 + fielding * 0.25;
  }
  return fielding;
}
